gulp-tasks/publish-bundle.js

```diff
@@ -0,0 +1,211 @@
+  return tarball.compress(bundleDirectory, tarPath)
+  .then(() => {
+    return {
+      bundleDirectory,
+      tarPath,
+    };
+  });
+};
+
+const handleGithubAndCDNRelease = (tagName, gitBranch, release) => {
+  let releaseCreation = Promise.resolve();
+  if (!release) {
+    releaseCreation = github.repos.createRelease({
+      owner: GITHUB_OWNER,
+      repo: GITHUB_REPO,
+      tag_name: tagName,
+      draft: true,
+      name: tagName,
+    })
+    .then((response) => {
+      release = response.data;
+    });
+  }
+
+  return releaseCreation
+  .then(() => {
+    return downloadGitCommit(gitBranch);
+  })
+  .then((tagDownloadPath) => {
+    return buildProject(tagDownloadPath)
+    .then(() => {
+      return bundleProject(tagDownloadPath, tagName);
+    });
+  })
+  .then(({bundleDirectory, tarPath}) => {
+    return github.repos.uploadAsset({
+      owner: GITHUB_OWNER,
+      repo: GITHUB_REPO,
+      id: release.id,
+      filePath: tarPath,
+      name: path.basename(tarPath),
+      label: path.basename(tarPath),
+    })
+    .then(() => {
+      return uploadBundleToCDN(tagName, bundleDirectory);
+    });
+  });
+};
+
+const findTagsWithoutBundles = () => {
+  // Get all of the tags in the repo.
+  return github.repos.getTags({
+    owner: GITHUB_OWNER,
+    repo: GITHUB_REPO,
+  })
+  .then((tagsResponse) => {
+    // We only want tags that are v3.0.0 or above.
+    const tagsData = tagsResponse.data;
+    return tagsData.filter((tagData) => {
+      // The "-alpha" means semver will match against tags with prelease
+      // info (although the prelease tag must be a string >= 'alpha'
+      // alphabetically speaking).
+      // TODO: Change to v3.0.0 when v3 is launched.
+      return semver.gte(tagData.name, 'v3.0.0-alpha');
+    });
+  })
+  .then((allTags) => {
+    return github.repos.getReleases({
+      owner: GITHUB_OWNER,
+      repo: GITHUB_REPO,
+    })
+    .then((releasesData) => {
+      const allReleases = releasesData.data;
+      return allTags.map((tagData) => {
+        let matchingRelease = null;
+
+        allReleases.forEach((release) => {
+          if (release.tag_name === tagData.name) {
+            matchingRelease = release;
+          }
+        });
+
+        return {
+          tagData: tagData,
+          release: matchingRelease,
+        };
+      });
+    });
+  })
+  .then((tagsAndReleases) => {
+    return tagsAndReleases.filter((tagAndRelease) => {
+      const release = tagAndRelease.release;
+      if (release && release.assets.length) {
+        // If a tag has a release and there is an asset let's assume the
+        // the release is fine. Note: Github's source doesn't count as an
+        // asset
+        return false;
+      }
+
+      return true;
+    });
+  });
+};
+
+gulp.task('publish-bundle:generate-from-tags', () => {
+  return findTagsWithoutBundles()
+  .then((tagsAndReleases) => {
+    return tagsAndReleases.reduce((promiseChain, tagAndRelease) => {
```

Comment:
With `async`/`await`, you can use a `for ... of` loop to execute asynchronous code in the proper sequence. `.reduce()` isn't needed anymore!

https://stackoverflow.com/a/37576787/385997